Mathlib/Combinatorics/SimpleGraph/Acyclic.lean

```diff
@@ -361,6 +361,113 @@ lemma Connected.exists_isTree_le [Finite V] (h : G.Connected) : ∃ T ≤ G, IsT
   obtain ⟨T, hTG, hmin⟩ := {H : SimpleGraph V | H.Connected}.toFinite.exists_le_minimal h
   exact ⟨T, hTG, isTree_of_minimal_connected hmin⟩
 
+/--
+Adding an edge to an acyclic graph preserves acyclicity if the endpoints are not reachable.
+-/
+theorem isAcyclic_add_edge_iff_of_not_reachable (x y : V) (hxy : ¬ G.Reachable x y) :
+    (G ⊔ fromEdgeSet {s(x,y)}).IsAcyclic ↔ IsAcyclic G := by
+  refine ⟨fun h ↦ h.anti le_sup_left, fun hG ↦ ?_⟩
+  have x_neq_y : x ≠ y := fun c => (c ▸ hxy) (Reachable.refl y)
+  have h_add_remove : (G ⊔ fromEdgeSet {s(x,y)}) \ fromEdgeSet {s(x,y)} = G := by
+    simpa using fun h => hxy h.reachable
+  have h_bridge : (G ⊔ fromEdgeSet {s(x,y)}).IsBridge s(x,y) := by
+    simpa [isBridge_iff, x_neq_y, h_add_remove]
+  rw [isBridge_iff_adj_and_forall_cycle_notMem] at h_bridge
+  intro u c hc
+  let c' : G.Walk u u := Walk.transfer c G (fun e he ↦ by
+    have eneq : e ≠ s(x,y) := fun h => h_bridge.2 c hc (h ▸ he)
+    simpa [eneq] using Walk.edges_subset_edgeSet c he)
+  exact hG c' (Walk.IsCycle.transfer (qc := hc) ..)
+
+/--
+The reachability relation of a maximal acyclic subgraph agrees with that of the larger graph.
+-/
+lemma reachable_eq_of_maximal_isAcyclic (F : SimpleGraph V)
+    (h : Maximal (fun H => H ≤ G ∧ H.IsAcyclic) F) : F.Reachable = G.Reachable := by
+  simp only [Maximal, and_imp] at h
+  obtain ⟨hF, h⟩ := h
+  apply funext; intro u; apply funext; intro v
+  refine propext ⟨fun hr => hr.mono hF.1, ?_⟩
+  contrapose! h
+  obtain ⟨p⟩ := h.1
+  let s : Set V := F.connectedComponentMk u
+  have hus : u ∈ s := ConnectedComponent.connectedComponentMk_mem
+  have hvs : v ∉ s := h.2 ∘ (F.connectedComponentMk u).reachable_of_mem_supp hus
+  obtain ⟨⟨⟨u', v'⟩, huv⟩, _, hu, hv⟩ := p.exists_boundary_dart s hus hvs
+  let F' := (F ⊔ fromEdgeSet {s(u', v')})
+  suffices F'.IsAcyclic by
+    rw [le_iff_adj] at hF
+    refine ⟨F', ?_, this, le_sup_left, ?_⟩
+    · have : G.Adj v' u' := G.symm huv
+      simp only [sup_le_iff, le_iff_adj, fromEdgeSet_adj, Set.mem_singleton_iff, Sym2.eq,
+      Sym2.rel_iff', Prod.mk.injEq, Prod.swap_prod_mk, ne_eq, and_imp, F']
+      grind
+    · rw [le_iff_adj]
+      push_neg
+      refine ⟨u', v', ?_, ?_⟩
+      · simpa [F'] using Or.inr huv.ne
+      · intro hc
+        have : _ := ConnectedComponent.mem_supp_congr_adj (F.connectedComponentMk u) hc
+        grind
+  refine (isAcyclic_add_edge_iff_of_not_reachable u' v' ?_).mpr hF.2
+  intro hc
+  rw [←ConnectedComponent.eq] at hc
+  suffices F.connectedComponentMk u' = s by
+    exact (hc ▸ this ▸ hv) ConnectedComponent.connectedComponentMk_mem
+  simp_rw [s, SetLike.coe, ConnectedComponent.supp_inj, ←ConnectedComponent.mem_supp_iff]
+  grind
+
+/-- A subgraph is maximal acyclic iff its reachability relation agrees with the larger graph. -/
+theorem maximal_isAcyclic_iff_reachable_eq {F : SimpleGraph V} (hF : F ≤ G ∧ F.IsAcyclic) :
```

Comment:
We prefer to supply multiple named hypotheses in lemmas rather than a conjunction:
```suggestion
theorem maximal_isAcyclic_iff_reachable_eq {F : SimpleGraph V} (hF : F ≤ G) (hF' : F.IsAcyclic) :
```